File: worksgood-pi/src/viz-snapshot.ts
/**
 * viz-snapshot.ts — live graph data for the embedded VizView panel.
 *
 * The panel reads the work graph from the **existing daemon IPC socket** (the
 * same one-request/one-response JSON-line `IpcRequest`/`IpcResponse` protocol
 * the TUI family speaks over `<wg-dir>/service/daemon.sock`, see
 * `src/commands/service/mod.rs::send_request_to_socket_with_timeout`). The
 * only request it ever sends is the read-only `viz_snapshot` — the panel never
 * mutates graph state (lifecycle authority stays with workers/controller).
 *
 * Resolution order for the socket path:
 *   1. `WG_DAEMON_SOCKET` (what WG exports / the forward-compat contract).
 *   2. `<WG_DIR>/service/daemon.sock` (WG_DIR is the workgraph dir).
 *   3. `<cwd>/.wg/service/daemon.sock` (standalone console pi in a WG repo).
 *
 * If the daemon is unreachable (no socket, connect error, timeout, error
 * response) the caller falls back to `wg viz --all --no-tui` ASCII output —
 * the same `generate_ascii` rendering the TUI shows — so the panel degrades
 * instead of disappearing.
 *
 * Polling is deliberately **bounded**: a fixed interval with an in-flight
 * guard (no overlapping requests), a bounded per-request timeout, and
 * change-detection so identical snapshots never re-render.
 */

import { connect, type Socket } from "node:net";
import { existsSync } from "node:fs";
import { join } from "node:path";
import type { WgBackend, WgEnv } from "./wg-backend.js";

/** Log entry from the bounded per-task tail. */
export interface VizLogEntry {
  timestamp?: string;
  actor?: string | null;
  message: string;
}

/** Aggregate token usage, mirroring the graph's `TokenUsage` JSON shape. */
export interface VizTokenUsage {
  cost_usd?: number;
  input_tokens?: number;
  output_tokens?: number;
  cache_read_input_tokens?: number;
  cache_creation_input_tokens?: number;
  total_tokens?: number;
}

/** One task of the bounded snapshot projection (`service::viz_snapshot`). */
export interface VizTask {
  id: string;
  title: string;
  presentation?: string;
  status: string;
  assigned?: string | null;
  after?: string[];
  before?: string[];
  description_head?: string | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  last_interaction_at?: string | null;
  token_usage?: VizTokenUsage | null;
  retry_count?: number;
  failure_reason?: string | null;
  log_count?: number;
  log_tail?: VizLogEntry[];
}

export interface VizSnapshot {
  tasks: VizTask[];
}

/** Where the resolved socket path came from (diagnostics only). */
export interface SocketResolution {
  socket: string | null;
  source: "env" | "wg-dir" | "cwd" | "none";
}

/**
 * Ordered daemon-socket candidates. WG_DIR may be either the workgraph dir
 * itself (`<project>/.wg`) or the project root depending on the launch path,
 * so both spellings are probed, then the cwd walk-below candidates.
 */
export function socketCandidates(
  env: Pick<WgEnv, "daemonSocket" | "dir">,
  cwd: string = process.cwd(),
): string[] {
  const candidates: string[] = [];
  if (env.daemonSocket) candidates.push(env.daemonSocket);
  if (env.dir) {
    candidates.push(join(env.dir, "service", "daemon.sock"));
    candidates.push(join(env.dir, ".wg", "service", "daemon.sock"));
  }
  candidates.push(join(cwd, ".wg", "service", "daemon.sock"));
  candidates.push(join(cwd, "service", "daemon.sock"));
  return candidates;
}

/** Resolve the first existing daemon socket path, or null when offline. */
export function resolveSocketPath(
  env: Pick<WgEnv, "daemonSocket" | "dir">,
  cwd: string = process.cwd(),
  exists: (p: string) => boolean = existsSync,
): SocketResolution {
  if (env.daemonSocket) return { socket: env.daemonSocket, source: "env" };
  if (env.dir) {
    const wgDir = join(env.dir, "service", "daemon.sock");
    if (exists(wgDir)) return { socket: wgDir, source: "wg-dir" };
    const projectRoot = join(env.dir, ".wg", "service", "daemon.sock");
    if (exists(projectRoot)) return { socket: projectRoot, source: "wg-dir" };
  }
  const cwdCandidate = join(cwd, ".wg", "service", "daemon.sock");
  if (exists(cwdCandidate)) return { socket: cwdCandidate, source: "cwd" };
  const legacy = join(cwd, "service", "daemon.sock");
  if (exists(legacy)) return { socket: legacy, source: "cwd" };
  return { socket: null, source: "none" };
}

export interface FetchOptions {
  /** Per-request deadline. The Rust client uses 2s for IPC; we match it. */
  timeoutMs?: number;
  /** Bounded log tail requested per task (clamped server-side to 1..=100). */
  logTail?: number;
  signal?: AbortSignal;
}

/**
 * One-shot daemon IPC round trip: connect, write `{"cmd":"viz_snapshot"…}\n`,
 * read the single JSON response line, close. Mirrors the Rust client's
 * one-request-per-connection contract exactly.
 */
export function fetchVizSnapshot(socketPath: string, opts: FetchOptions = {}): Promise<VizSnapshot> {
  const timeoutMs = opts.timeoutMs ?? 2000;
  const request = JSON.stringify({ cmd: "viz_snapshot", log_tail: opts.logTail ?? 20 });
  return new Promise<VizSnapshot>((resolve, reject) => {
    let settled = false;
    let buffer = "";
    let socket: Socket | null = null;
    const finish = (err: Error | null, value?: VizSnapshot) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      if (opts.signal) opts.signal.removeEventListener("abort", onAbort);
      try {
        socket?.destroy();
      } catch {
        /* socket already gone */
      }
      if (err) reject(err);
      else resolve(value as VizSnapshot);
    };
    const timer = setTimeout(() => finish(new Error(`viz_snapshot timed out after ${timeoutMs}ms`)), timeoutMs);
    const onAbort = () => finish(new Error("viz_snapshot aborted"));
    if (opts.signal) {
      if (opts.signal.aborted) return finish(new Error("viz_snapshot aborted"));
      opts.signal.addEventListener("abort", onAbort, { once: true });
    }

    socket = connect(socketPath, () => {
      socket?.write(`${request}\n`);
    });
    socket.once("error", (err: Error) => finish(err));
    socket.on("data", (chunk: Buffer) => {
      buffer += chunk.toString("utf8");
      let idx = buffer.indexOf("\n");
      while (idx >= 0) {
        const line = buffer.slice(0, idx).trim();
        buffer = buffer.slice(idx + 1);
        if (line) {
          try {
            const response = JSON.parse(line) as { ok?: boolean; error?: string } & Partial<VizSnapshot>;
            if (response.ok === false) {
              return finish(new Error(response.error ?? "viz_snapshot failed"));
            }
            const tasks = (response as VizSnapshot).tasks;
            if (!Array.isArray(tasks)) {
              return finish(new Error("viz_snapshot response missing tasks"));
            }
            return finish(null, { tasks });
          } catch (err) {
            return finish(err instanceof Error ? err : new Error(String(err)));
          }
        }
        idx = buffer.indexOf("\n");
      }
    });
    socket.once("close", () => {
      if (!settled && !buffer.trim()) {
        finish(new Error("daemon closed the connection without a response"));
      }
    });
  });
}

export type PollCallback = (snapshot: VizSnapshot) => void;

/**
 * Bounded poller: fixed interval, no overlapping in-flight requests, and a
 * change guard so identical snapshots are never re-delivered.
 */
export class VizPoller {
  private timer: ReturnType<typeof setInterval> | null = null;
  private inFlight = false;
  private lastJson: string | null = null;
  private stopped = true;

  constructor(
    private readonly fetcher: () => Promise<VizSnapshot>,
    private readonly onChange: PollCallback,
    private readonly intervalMs = 5000,
  ) {}

  start(): void {
    if (!this.stopped) return;
    this.stopped = false;
    this.timer = setInterval(() => void this.refresh(), this.intervalMs);
    void this.refresh();
  }

  stop(): void {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  /** One immediate bounded fetch; safe to call while a fetch is in flight. */
  async refresh(): Promise<VizSnapshot | null> {
    if (this.inFlight || this.stopped) return null;
    this.inFlight = true;
    try {
      const snapshot = await this.fetcher();
      const json = JSON.stringify(snapshot);
      if (json !== this.lastJson) {
        this.lastJson = json;
        this.onChange(snapshot);
      }
      return snapshot;
    } catch {
      // Silent degrade: an offline daemon is a normal state (the panel falls
      // back to ASCII), never an error surface inside pi.
      return null;
    } finally {
      this.inFlight = false;
    }
  }
}

export type VizDataSource =
  | { kind: "snapshot"; snapshot: VizSnapshot; source: string }
  | { kind: "ascii"; text: string };

/**
 * Snapshot-with-fallback: try the daemon socket; on any failure (offline,
 * timeout, error response) shell `wg viz --all --no-tui` and surface the same
 * ASCII the static viz prints. The fallback is a read-only verb.
 */
export async function vizSnapshotWithFallback(
  backend: Pick<WgBackend, "run">,
  env: WgEnv,
  opts: FetchOptions = {},
): Promise<VizDataSource> {
  const resolved = resolveSocketPath(env);
  if (resolved.socket) {
    try {
      const snapshot = await fetchVizSnapshot(resolved.socket, opts);
      return { kind: "snapshot", snapshot, source: resolved.source };
    } catch {
      // fall through to the ASCII fallback
    }
  }
  const r = await backend.run(["viz", "--all", "--no-tui"], { signal: opts.signal });
  return { kind: "ascii", text: (r.stdout || r.stderr || "").trimEnd() };
}

File: src/service/viz_snapshot.rs
//! Read-only graph snapshot projection for embedded graph viewers.
//!
//! The pi plugin (`worksgood-pi`) renders a live work-graph panel inside a Pi
//! session. Rather than reading `graph.jsonl` from disk (a second, drifting
//! reader) it polls the daemon's IPC lane — the same protocol the TUI family
//! uses — with a `VizSnapshot` request served by
//! `crate::commands::service::ipc::handle_viz_snapshot`, which delegates the
//! per-task projection to [`viz_snapshot_task`].
//!
//! The projection is deliberately **bounded**: identity, status, dependency
//! edges, timing, aggregate token usage, and a small log tail. No
//! descriptions, transcripts, receipts, or completion evidence — a viewer
//! surface never becomes a bulk channel over the socket, and the request never
//! mutates graph or control-plane state (lifecycle authority stays with
//! workers and the coordinator).

use crate::graph::Task;
use serde_json::Value;

/// Default per-task log entries included in a snapshot.
pub const DEFAULT_LOG_TAIL: usize = 20;

/// Clamp a client-supplied log-tail request into the bounded range.
pub fn clamp_log_tail(requested: Option<usize>) -> usize {
    requested.unwrap_or(DEFAULT_LOG_TAIL).clamp(1, 100)
}

/// Project one task into the bounded JSON shape the viz panel consumes.
///
/// Field names are kebab-case and mirror the graph's own serde encoding so a
/// viewer reads one vocabulary (`in-progress`, `done`, …) instead of a
/// parallel debug-format spelling.
pub fn viz_snapshot_task(task: &Task, log_tail: usize) -> Value {
    let log_tail_entries: Vec<Value> = task
        .log
        .iter()
        .rev()
        .take(log_tail)
        .map(|entry| {
            serde_json::json!({
                "timestamp": entry.timestamp,
                "actor": entry.actor,
                "message": entry.message,
            })
        })
        .collect();
    serde_json::json!({
        "id": task.id,
        "title": task.title,
        "presentation": task.presentation,
        "status": task.status,
        "assigned": task.assigned,
        "after": task.after,
        "before": task.before,
        // Bounded head of the description (the TUI HUD detail shows the full
        // text; the panel gets a bounded preview). Truncated on a char
        // boundary at 512 bytes with an ellipsis marker.
        "description_head": task.description.as_deref().map(|d| {
            if d.len() <= 512 {
                d.to_string()
            } else {
                let mut cut = 512;
                while cut > 0 && !d.is_char_boundary(cut) {
                    cut -= 1;
                }
                format!("{}…", &d[..cut])
            }
        }),
        "created_at": task.created_at,
        "started_at": task.started_at,
        "completed_at": task.completed_at,
        "last_interaction_at": task.last_interaction_at,
        "token_usage": task.token_usage,
        "retry_count": task.retry_count,
        "failure_reason": task.failure_reason,
        "log_count": task.log.len(),
        "log_tail": log_tail_entries,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graph::{LogEntry, Node, Status, WorkGraph};
    use crate::test_helpers::make_task_with_status;

    #[test]
    fn viz_snapshot_projection_is_bounded_and_read_only() {
        let mut graph = WorkGraph::new();
        let mut parent = make_task_with_status("parent-a", "Parent A", Status::Done);
        for i in 0..30 {
            parent.log.push(LogEntry {
                timestamp: format!("2026-01-01T00:00:{:02}Z", i),
                actor: Some("tester".to_string()),
                user: None,
                message: format!("entry {i}"),
            });
        }
        let mut child = make_task_with_status("child-b", "Child B", Status::InProgress);
        child.after = vec!["parent-a".to_string()];
        child.assigned = Some("agent-1".to_string());
        for i in 0..30 {
            child.log.push(LogEntry {
                timestamp: format!("2026-01-01T00:00:{:02}Z", i),
                actor: Some("tester".to_string()),
                user: None,
                message: format!("entry {i}"),
            });
        }
        graph.add_node(Node::Task(parent));
        graph.add_node(Node::Task(child));

        let tail = clamp_log_tail(None);
        assert_eq!(tail, 20);
        assert_eq!(clamp_log_tail(Some(0)), 1, "clamped to the 1..=100 range");
        assert_eq!(clamp_log_tail(Some(500)), 100);

        let projected = graph
            .tasks()
            .map(|t| viz_snapshot_task(t, tail))
            .collect::<Vec<_>>();
        assert_eq!(projected.len(), 2);

        let child_json = projected.iter().find(|t| t["id"] == "child-b").unwrap();
        assert_eq!(child_json["status"], "in-progress");
        assert_eq!(child_json["presentation"], "primary");
        assert_eq!(child_json["after"], serde_json::json!(["parent-a"]));
        assert_eq!(child_json["assigned"], "agent-1");
        // The log tail is bounded even when the task carries 30 entries, and
        // the tail is newest-first.
        assert_eq!(child_json["log_count"], 30);
        let entries = child_json["log_tail"].as_array().unwrap();
        assert_eq!(entries.len(), 20);
        assert_eq!(entries[0]["message"], "entry 29");
        assert_eq!(entries[19]["message"], "entry 10");
        // No description / transcript fields leak into the bounded projection.
        assert!(child_json.get("description").is_none());
    }
}
